### PR TITLE
Better app password support for bitbucket

### LINKED ISSUE #12102: If a bitbucket repo is listed in repositories with ssh url, authentication will require ssh key
<!--
The information we ask for below is IMPORTANT. It will help us help you faster.
Please FOLLOW THE ISSUE TEMPLATE unless you have a good reason not to, and help avoid wasting everyone's time.

Please also make sure you are using the latest Composer version. You can get the latest with `composer self-update`
or even try the latest dev build with `composer self-update --snapshot` to see if the bug was perhaps already fixed but not released yet.
-->

My `composer.json`:

```json
{
  "name": "my-vendor/my-package",
  "repositories": [
    {
      "type": "vcs",
      "url": "git@bitbucket.org:user/test-dep.git"
    }
  ],
  "minimum-stability": "dev",
  "require": {
    "php": ">=5.3.0",
    "user/test-dep": "dev-main"
  }
}

```

Output of `composer diagnose`:

```
Composer could not detect the root package (my-vendor/my-package) version, defaulting to '1.0.0'. See https://getcomposer.org/root-version
Checking composer.json: WARNING
No license specified, it is recommended to do so. For closed-source software you may use "proprietary" as license.
Checking platform settings: OK
Checking git settings: OK git version 2.45.2
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK does not expire
Checking disk free space: OK
Checking pubkeys: FAIL
Missing pubkey for tags verification
Missing pubkey for dev verification
Run composer self-update --update-keys to set them up
Checking Composer version: You are not running the latest stable version, run `composer self-update` to update (2.7.8 => 2.7.9)
Checking Composer and its dependencies for vulnerabilities: OK
Composer version: 2.7.8
PHP version: 8.1.29
PHP binary path: /usr/local/bin/php
OpenSSL version: OpenSSL 3.3.1 4 Jun 2024
curl version: 8.8.0 libz 1.3.1 ssl OpenSSL/3.3.1
zip: extension present, unzip present, 7-Zip not available

```

When I run this command: <!-- run it with `-vvv` added to it ideally to get full debug output -->

```
composer update -vvv
```

I get the following output: <!-- FULL OUTPUT please, not just what you think is relevant -->

```
Running 2.7.8 (2024-08-22 15:28:36) with PHP 8.1.29 on Linux / 6.8.0-40-generic
Reading ./composer.json (/derp/composer.json)
Loading config file ./composer.json (/derp/composer.json)
Loading config file /derp/auth.json
Reading /derp/auth.json
Checked CA file /etc/pki/tls/certs/ca-bundle.crt does not exist or it is not a file.
Checked directory /etc/pki/tls/certs/ca-bundle.crt does not exist or it is not a directory.
Checked CA file /etc/ssl/certs/ca-certificates.crt: valid
Executing command (/derp): 'git' 'branch' '-a' '--no-color' '--no-abbrev' '-v'
Executing command (/derp): git describe --exact-match --tags
Executing command (CWD): git --version
Executing command (/derp): git log --pretty="%H" -n1 HEAD --no-show-signature
Executing command (/derp): hg branch
Executing command (/derp): fossil branch list
Executing command (/derp): fossil tag list
Executing command (/derp): svn info --xml
Composer could not detect the root package (my-vendor/my-package) version, defaulting to '1.0.0'. See https://getcomposer.org/root-version
Failed to initialize global composer: Composer could not find the config file: /root/.composer/composer.json

Reading ./composer.lock (/derp/composer.lock)
Loading composer repositories with package information
Executing command (CWD): git clone --mirror -- 'git@bitbucket.org:user/test-dep.git' '/root/.composer/cache/vcs/git-bitbucket.org-user-test-dep.git/'
Executing command (CWD): git --version

In Git.php line 478:
                                                                                                                                                  
  [RuntimeException]                                                                                                                              
  Failed to execute git clone --mirror -- 'git@bitbucket.org:user/test-dep.git' '/root/.composer/cache/vcs/git-bitbucket.org-user-test-dep.git/'  
                                                                                                                                                  
  Cloning into bare repository '/root/.composer/cache/vcs/git-bitbucket.org-user-test-dep.git'...                                                 
  git@bitbucket.org: Permission denied (publickey).                                                                                               
  fatal: Could not read from remote repository.                                                                                                   
                                                                                                                                                  
  Please make sure you have the correct access rights                                                                                             
  and the repository exists.                                                                                                                      
                                                                                                                                                  

Exception trace:
  at phar:///usr/local/bin/composer/src/Composer/Util/Git.php:478
 Composer\Util\Git->throwException() at phar:///usr/local/bin/composer/src/Composer/Util/Git.php:265
 Composer\Util\Git->runCommand() at phar:///usr/local/bin/composer/src/Composer/Util/Git.php:302
 Composer\Util\Git->syncMirror() at phar:///usr/local/bin/composer/src/Composer/Repository/Vcs/GitDriver.php:71
 Composer\Repository\Vcs\GitDriver->initialize() at phar:///usr/local/bin/composer/src/Composer/Repository/VcsRepository.php:144
 Composer\Repository\VcsRepository->getDriver() at phar:///usr/local/bin/composer/src/Composer/Repository/VcsRepository.php:190
 Composer\Repository\VcsRepository->initialize() at phar:///usr/local/bin/composer/src/Composer/Repository/ArrayRepository.php:308
 Composer\Repository\ArrayRepository->getPackages() at phar:///usr/local/bin/composer/src/Composer/Repository/ArrayRepository.php:62
 Composer\Repository\ArrayRepository->loadPackages() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/PoolBuilder.php:425
 Composer\DependencyResolver\PoolBuilder->loadPackagesMarkedForLoading() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/PoolBuilder.php:275
 Composer\DependencyResolver\PoolBuilder->buildPool() at phar:///usr/local/bin/composer/src/Composer/Repository/RepositorySet.php:332
 Composer\Repository\RepositorySet->createPool() at phar:///usr/local/bin/composer/src/Composer/Installer.php:501
 Composer\Installer->doUpdate() at phar:///usr/local/bin/composer/src/Composer/Installer.php:298
 Composer\Installer->run() at phar:///usr/local/bin/composer/src/Composer/Command/UpdateCommand.php:251
 Composer\Command\UpdateCommand->execute() at phar:///usr/local/bin/composer/vendor/symfony/console/Command/Command.php:298
 Symfony\Component\Console\Command\Command->run() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:1040
 Symfony\Component\Console\Application->doRunCommand() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:301
 Symfony\Component\Console\Application->doRun() at phar:///usr/local/bin/composer/src/Composer/Console/Application.php:390
 Composer\Console\Application->doRun() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:171
 Symfony\Component\Console\Application->run() at phar:///usr/local/bin/composer/src/Composer/Console/Application.php:148
 Composer\Console\Application->run() at phar:///usr/local/bin/composer/bin/composer:93
 require() at /usr/local/bin/composer:29

update [--with WITH] [--prefer-source] [--prefer-dist] [--prefer-install PREFER-INSTALL] [--dry-run] [--dev] [--no-dev] [--lock] [--no-install] [--no-audit] [--audit-format AUDIT-FORMAT] [--no-autoloader] [--no-suggest] [--no-progress] [-w|--with-dependencies] [-W|--with-all-dependencies] [-v|vv|vvv|--verbose] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--apcu-autoloader] [--apcu-autoloader-prefix APCU-AUTOLOADER-PREFIX] [--ignore-platform-req IGNORE-PLATFORM-REQ] [--ignore-platform-reqs] [--prefer-stable] [--prefer-lowest] [-m|--minimal-changes] [-i|--interactive] [--root-reqs] [--] [<packages>...]


```

And I expected this to happen:

Be able to clone, since I have an app password stored as basic auth for bitbucket.org

If I do the same notation for a VCS repo for gitlab or github, I am able to clone it fine.

### PR DESCRIPTION
**Pull request by eiriksm**
This fixes #12102 

It also fixes what is kind of a bad UX if you have stored actual credentials for bitbucket (for example an app password) and composer will then try to first request an access token for you using them, and in the process show some errors:

```
./composer/bin/composer update
Composer is operating slower than normal because you have Xdebug enabled. See https://getcomposer.org/xdebug
Composer could not detect the root package (my-vendor/my-package) version, defaulting to '1.0.0'. See https://getcomposer.org/root-version
Loading composer repositories with package information
Invalid OAuth consumer provided.
This can have three reasons:
1. You are authenticating with a bitbucket username/password combination
2. You are using an OAuth consumer, but didn't configure a (dummy) callback url
3. You are using an OAuth consumer, but didn't configure it as private consumer
Invalid OAuth consumer provided.
This can have three reasons:
1. You are authenticating with a bitbucket username/password combination
2. You are using an OAuth consumer, but didn't configure a (dummy) callback url
3. You are using an OAuth consumer, but didn't configure it as private consumer
Updating dependencies                                                          
Nothing to modify in lock file
Installing dependencies from lock file (including require-dev)
Package operations: 2 installs, 0 updates, 0 removals
  - Syncing eiriksm/test-dep-bitbucket (dev-main 9cf2753) into cache
  - Downloading eiriksm/test-dep-github (dev-main bd6bc34)
  - Installing eiriksm/test-dep-bitbucket (dev-main 9cf2753): Cloning 9cf275369a from cache
  - Installing eiriksm/test-dep-github (dev-main bd6bc34): Extracting archive
Generating autoload files
No security vulnerability advisories found.
```

There are actually colors and everything. Looks super scary and wrong. And 2 times even. But it's actually just warnings, and the code path that works, is executed later. Attached also an image of this scary warning

Screenshot from 2024-09-08 21-04-45

So yeah this PR fixes both of those issues. I only created an actual issue for the first of them. I could go ahead and create one for the other one if appropriate and needed for bookkeeping?

### CODE REVIEW

File: src/Composer/Util/Git.php [hunk 1]
```diff
@@ -131,42 +131,59 @@ public function runCommand(callable $commandCallable, string $url, ?string $cwd,
...
+            ) { //bitbucket either through oauth or app password, with fallback to ssh.
+                if ($match[1] === 'git') {
+                    $match[1] = 'https';
+                }
                 $bitbucketUtil = new Bitbucket($this->io, $this->config, $this->process);
 
-                if (!$this->io->hasAuthentication($match[1])) {
+                $domain = $match[2];
+
+                // First we try to athenticate with whatever we have stored.
```
Comment (stof) [inline src/Composer/Util/Git.php:145]:
```suggestion
                // First we try to authenticate with whatever we have stored.
```

File: src/Composer/Util/Git.php [hunk 2]
```diff
@@ -131,42 +131,59 @@ public function runCommand(callable $commandCallable, string $url, ?string $cwd,
                     $errorMsg = $this->process->getErrorOutput();
                 }
             // @phpstan-ignore composerPcre.maybeUnsafeStrictGroups
-            } elseif (Preg::isMatchStrictGroups('{^https://(bitbucket\.org)/(.*?)(?:\.git)?$}i', $url, $match)) { //bitbucket oauth
+            } elseif (
+                Preg::isMatchStrictGroups('{^(https?)://(bitbucket\.org)/(.*)}i', $url, $match)
```
Comment (stof) [inline src/Composer/Util/Git.php:135]: why is this changing the regex ?

Comment (eiriksm) [reply src/Composer/Util/Git.php:135]: Good question.

It's changing it for a couple reasons:

1. It's closer to the other branches of matching VCS providers
2. It makes it possible to use the same key for $matches regardless of if we use `https` or `git@` which seems to be the motivation to use a similar pattern on the other branches of matching (for example)

For example the refactoring I did to assign the domain variable from its corresponding $matches key would be mroe ugly if for one of the cases it would be $matches[1] and the other it would be $matches[2]

To conclude, I copy pasted from gitlab and basically just made appropriate edits for bitbucket. I could revert it of course, and adapt the git@ regex to get the matches array similar?

Comment (stof) [reply src/Composer/Util/Git.php:135]: but this changes what `$matches[2]` contains for an HTTPS url using the `.git` suffix. Wouldn't this cause issues ?

Comment (eiriksm) [reply src/Composer/Util/Git.php]: I see what you mean.

I cleaned it up a bit, and added tests. So I am at least no longer wondering if it works with or without the .git part :smile:

File: src/Composer/Util/Git.php [hunk 3]
```diff
@@ -131,42 +131,60 @@ public function runCommand(callable $commandCallable, string $url, ?string $cwd,
...
+                        return;
+                    }
+                    $credentials = [
+                        rawurlencode($auth['username']),
+                        rawurlencode($auth['password'])
+                    ];
+                    $errorMsg = $this->process->getErrorOutput();
+                }
+
+                if (!$this->io->hasAuthentication($domain)) {
```
Comment (Seldaek) [inline src/Composer/Util/Git.php:166]: Why not first do the if !hasAuthentication branch? And then once/if we have auth, we try to call $commandCallable() with that auth, and if that does not work you could do the fallback to ssh then? Right now you first try with some auth, but then prompt for auth, and then that doesn't really get tried as far as I can tell?

Comment (eiriksm) [reply src/Composer/Util/Git.php:166]: I wrote a longer answer here, but now I understand what you mean. I believe you may be totally right in that, let me have a look at that

Comment (eiriksm) [reply src/Composer/Util/Git.php]: Refactored a bit, and wrote tests for that part as well

**Comment by eiriksm** [comment]
Updated PR, I see there is a failing phpstan run on PHP 8.3 but seems unrelated to this PR

**Comment by Seldaek** [comment]
Thanks, I just hope this won't cause regressions for some edge cases.

**Comment by eiriksm** [comment]
I hope so too 😬

At least now we have the test coverage in place and it would be much more trivial to refactor and add more tests

**Comment by Seldaek** [comment]
@eiriksm could you please take a look at https://github.com/composer/composer/issues/12441 see if it makes sense to you?

**Comment by eiriksm** [comment]
Hey there. Just wanted to say I saw the message and I hope I will be able to. Just has not been possible yet 😅

**Comment by Seldaek** [comment]
No worries, it's fixed by https://github.com/composer/composer/pull/12462
